pkg/bucketaws/setup.go

```diff
@@ -18,64 +18,131 @@ package s3setup
+		year, month, day := d.Date()
+		dates = append(
+			dates,
+			fmt.Sprintf(
+				"created_year=%d/created_month=%d/created_day=%d",
+				year, month, day))
+	}
+	return dates
+}
+
 func ConfigRead() error {
+
 	// Setting up file formatting
-	// Pulling from Viper
-	viper.SetConfigFile("config/syncron.yaml")
+	// Using Viper
+	// Reading from file syncron.yaml
+
+	viper.AddConfigPath("./config")
+	viper.SetConfigFile("syncron.yaml")
 	viper.SetConfigType("yaml")
-	viper.AddConfigPath(".")
+	viper.SetDefault("download_dir", "/tmp/syncron/")
+
 	// Reading from file
 	err := viper.ReadInConfig()
 	if err != nil {
-		return err
-	} else {
-		logrus.Info("Your configuration file was read succesfully")
-		logrus.Info("Reading from bucket: ", viper.Get("bucket"))
+		logrus.Error(err)
+		logrus.Exit(1)
 	}
+	logrus.Info("Your configuration file was read succesfully")
+	logrus.Info("Reading from bucket: ", viper.Get("bucket"))
 	return nil
 }
 
 func SetupSession() *session.Session {
+
 	// Initialize a session with AWS SDK
-	// It will read from the file located at ~/.aws/credentials and syncron/config/syncron.yaml
+	// It will read from the file located at ~/.aws/credentials and syncron/syncron.yaml
+
 	sess, err := session.NewSession(&aws.Config{
 		Region:   aws.String(viper.GetString("s3.region")),
 		Endpoint: aws.String(viper.GetString("s3.endpoint")),
 	},
 	)
 	if err != nil {
-		fmt.Println("There was an error setting up your aws session")
+		fmt.Println("There was an error setting up your aws session", err)
 		os.Exit(1)
-	} else {
-		logrus.Info("Your AWS session was set up correctly")
 	}
+	logrus.Info("Your AWS session was set up correctly")
 	return sess
 }
 
-func AccessBucket(sess *session.Session) error {
+func AccessBucket(sess *session.Session) (*s3.S3, *s3manager.Downloader) {
+
+	// This function initiates the service for downloading files in s3
+
 	logrus.Info("Accessing bucket...")
 	svc := s3.New(sess)
-	resp, err := svc.ListObjectsV2(&s3.ListObjectsV2Input{Bucket: aws.String(viper.GetString("bucket"))})
-	if err != nil {
-		fmt.Println("Unable to list items in bucket", resp)
-		return err
-	} else {
-		logrus.Info("Success getting into the bucket!")
-	}
-	return nil
+	dwn := s3manager.NewDownloader(sess)
+
+	return svc, dwn
 }
 
-func DownloadFromBucket() error {
+func DownloadFromBucket(svc *s3.S3, dwn *s3manager.Downloader, dates []string, bprefix string) error {
 
-	// Future PR implementation
+	// This function takes care of listing the keys in the bucket, filtering
+	// through those that are needed.
 
+	var continuationToken *string
+
+	for {
+		resp, err := svc.ListObjectsV2(&s3.ListObjectsV2Input{
+			Bucket:            aws.String(viper.GetString("bucket")),
+			Prefix:            aws.String(files.AppendPrefix(bprefix)),
+			ContinuationToken: continuationToken,
+		})
+		if err != nil {
+			logrus.Error("There was an error listing the objects in bucket.")
+			logrus.Error(err)
+			os.Exit(1)
```

Comment:
I am not fond of crashing the application against an error. What if as a caller I wanted to try again with another configuration? It is unexpected behavior that makes things harder to work with.

Reply:
This was my way of interrupting the execution, as it was throwing the error message, but the application was continuing.
Any suggestions for a better implementation?

Reply:
I think that what @jsanemet meant in this comment is not so much about `Fatal` or `Error`+`Exit` (according to the documentation `Fatal` does internally already call `Exit`, so it's essentially the same), but rather that maybe the function should return the error and then it should be the caller who should decide whether to throw or not. I agree with him, I would not expect a call to this function to interrupt the program, but it depends on how it's intended to be called.